server/src/main/java/io/crate/planner/operators/JoinPlan.java

```diff
@@ -21,79 +21,126 @@
+                               PlannerContext plannerContext,
+                               Set<PlanHint> planHints,
+                               ProjectionBuilder projectionBuilder,
+                               int limit,
+                               int offset,
+                               @Nullable OrderBy order,
+                               @Nullable Integer pageSizeHint,
+                               Row params,
+                               SubQueryResults subQueryResults) {
+        throw new UnsupportedOperationException("JoinPlan can not be build, convert to Hash-join/NestedLoop Join instead");
     }
 
-    @Nullable
-    public Symbol joinCondition() {
-        return joinCondition;
+    @Override
+    public <C, R> R accept(LogicalPlanVisitor<C, R> visitor, C context) {
+        return visitor.visitJoinPlan(this, context);
     }
 
-    public JoinType joinType() {
-        return joinType;
+    public boolean crossJoinEliminationRulePassed() {
+        return crossJoinEliminationRulePassed;
     }
 
 
-    protected static MergePhase buildMergePhaseForJoin(PlannerContext plannerContext,
-                                             ResultDescription resultDescription,
-                                             Collection<String> executionNodes) {
-        List<Projection> projections = Collections.emptyList();
-        if (resultDescription.hasRemainingLimitOrOffset()) {
-            projections = Collections.singletonList(ProjectionBuilder.limitAndOffsetOrEvalIfNeeded(
-                resultDescription.limit(),
-                resultDescription.offset(),
-                resultDescription.numOutputs(),
-                resultDescription.streamOutputs()
-            ));
+    @Override
+    public LogicalPlan pruneOutputsExcept(Collection<Symbol> outputsToKeep) {
+        LinkedHashSet<Symbol> lhsToKeep = new LinkedHashSet<>();
+        LinkedHashSet<Symbol> rhsToKeep = new LinkedHashSet<>();
+        for (Symbol outputToKeep : outputsToKeep) {
+            SymbolVisitors.intersection(outputToKeep, lhs.outputs(), lhsToKeep::add);
+            SymbolVisitors.intersection(outputToKeep, rhs.outputs(), rhsToKeep::add);
         }
+        if (joinCondition != null) {
+            SymbolVisitors.intersection(joinCondition, lhs.outputs(), lhsToKeep::add);
+            SymbolVisitors.intersection(joinCondition, rhs.outputs(), rhsToKeep::add);
+        }
+        LogicalPlan newLhs = lhs.pruneOutputsExcept(lhsToKeep);
+        LogicalPlan newRhs = rhs.pruneOutputsExcept(rhsToKeep);
+        if (newLhs == lhs && newRhs == rhs) {
+            return this;
+        }
+        return new JoinPlan(
+            newLhs,
+            newRhs,
+            joinType,
+            joinCondition,
+            isFiltered,
+            crossJoinEliminationRulePassed
+        );
+    }
+
+    @Override
+    public void print(PrintContext printContext) {
+        printContext
+            .text("JoinPlan[")
```

Comment:
```suggestion
            .text("Join[")
```

The `Plan` is a bit redundant